macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift

```diff
@@ -2057,7 +2113,61 @@ extension AddressBarButtonsViewController {
 
 extension AddressBarButtonsViewController: NavigationBarBadgeAnimatorDelegate {
 
-    func didFinishAnimating() {
+    func didFinishAnimating(type: NavigationBarBadgeAnimationView.AnimationType) {
+        // If a tracker notification just finished, play the shield Lottie animation (HTTPS only)
+        if case .trackersBlocked = type {
+            // Check if we have a valid URL to play shield animation
+            guard let tabViewModel = tabViewModel,
+                  case .url(let url, _, _) = tabViewModel.tab.content else {
+                // Tab navigated away from URL content - still need to process queue
+                buttonsBadgeAnimator.processNextAnimation()
+                playPrivacyInfoHighlightAnimationIfNecessary()
+                return
+            }
+
+            // Check if animator is busy before starting shield animation
+            guard !buttonsBadgeAnimator.isAnimating else {
+                playPrivacyInfoHighlightAnimationIfNecessary()
+                return
+            }
+
+            // Only play shield animation for HTTPS sites
+            // HTTP and unprotected sites use static icons instead
+            guard url.navigationalScheme != .http else {
+                buttonsBadgeAnimator.processNextAnimation()
+                playPrivacyInfoHighlightAnimationIfNecessary()
+                return
+            }
+
+            // Capture URL at animation start for validation in completion handler
+            let animationURL = url
+
+            // Play the Lottie animation from frame 1 to the end
+            // The animation is already visible at frame 1 as the privacy button icon
+            let endFrame = shieldAnimationView.animation?.endFrame ?? 0
+            shieldAnimationView.play(fromFrame: 1, toFrame: endFrame, loopMode: .playOnce) { [weak self] finished in
+                // Stop at the last frame when animation completes (don't rewind or loop)
+                guard finished, let self = self else { return }
+
+                // Verify we're still on the same page before updating state
+                guard case .url(let currentURL, _, _) = self.tabViewModel?.tab.content,
+                      currentURL == animationURL else { return }
+
+                self.shieldAnimationView.pause()
+                self.shieldAnimationView.currentFrame = endFrame
+                // Mark that shield animation has completed to prevent reset
+                self.hasShieldAnimationCompleted = true
+
+                // After shield animation completes, process next queued notification (like cookies)
+                self.buttonsBadgeAnimator.processNextAnimation()
+                self.playPrivacyInfoHighlightAnimationIfNecessary()
+            }
+
+            // Return early - don't process next animation yet, wait for shield to complete
+            return
+        }
```

Comment:
I have a hard time understanding this piece of code - it's long, quite heavy in logic - e.g. there are many returns and some without queue not advancing, I'm not sure if this is valid or not?

Could you also extract code that runs the animation into a separate `func playShieldAnimation(for url: URL)` not to combine animation code with our business logic?

Reply:
Yea, the early returns are intentional (no URL, animator busy, not Https site), which should not process the animation.

Good call on the method.  Extracted.